Endless snapshot restore loop

After restarting a Node it fails to rejoin because of a endless loop of restoring.
I think the problem is that this snapshot already exist at this node and all chunks are ignored, which means the temp directory is not created. Since the temp directory is not created this will fail on moving the temp snapshot. 

The restore will completely fail even if the node has this snapshot already.

```
09:59:59.432 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000108.sst, because snapshot 8894878975848 already exists.
09:59:59.433 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000109.sst, because snapshot 8894878975848 already exists.
09:59:59.434 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000111.sst, because snapshot 8894878975848 already exists.
09:59:59.435 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000112.sst, because snapshot 8894878975848 already exists.
09:59:59.436 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000113.sst, because snapshot 8894878975848 already exists.
09:59:59.440 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000114.sst, because snapshot 8894878975848 already exists.
09:59:59.443 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000117.sst, because snapshot 8894878975848 already exists.
09:59:59.445 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000121.sst, because snapshot 8894878975848 already exists.
09:59:59.446 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000122.sst, because snapshot 8894878975848 already exists.
09:59:59.448 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000123.sst, because snapshot 8894878975848 already exists.
09:59:59.449 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000124.sst, because snapshot 8894878975848 already exists.
09:59:59.452 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000125.sst, because snapshot 8894878975848 already exists.
09:59:59.453 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000126.sst, because snapshot 8894878975848 already exists.
09:59:59.454 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000152.sst, because snapshot 8894878975848 already exists.
09:59:59.454 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000154.sst, because snapshot 8894878975848 already exists.
09:59:59.458 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000156.sst, because snapshot 8894878975848 already exists.
09:59:59.461 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000162.sst, because snapshot 8894878975848 already exists.
09:59:59.463 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000168.sst, because snapshot 8894878975848 already exists.
09:59:59.464 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000170.sst, because snapshot 8894878975848 already exists.
09:59:59.466 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000172.sst, because snapshot 8894878975848 already exists.
09:59:59.467 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000174.sst, because snapshot 8894878975848 already exists.
09:59:59.471 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000176.sst, because snapshot 8894878975848 already exists.
09:59:59.472 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000177.sst, because snapshot 8894878975848 already exists.
09:59:59.473 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000182.sst, because snapshot 8894878975848 already exists.
09:59:59.475 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000185.sst, because snapshot 8894878975848 already exists.
09:59:59.550 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000188.sst, because snapshot 8894878975848 already exists.
09:59:59.551 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000190.sst, because snapshot 8894878975848 already exists.
09:59:59.552 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000191.sst, because snapshot 8894878975848 already exists.
09:59:59.554 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000192.sst, because snapshot 8894878975848 already exists.
09:59:59.554 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000193.sst, because snapshot 8894878975848 already exists.
09:59:59.555 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000194.sst, because snapshot 8894878975848 already exists.
09:59:59.558 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000195.sst, because snapshot 8894878975848 already exists.
09:59:59.560 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000196.sst, because snapshot 8894878975848 already exists.
09:59:59.562 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000197.sst, because snapshot 8894878975848 already exists.
09:59:59.757 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000198.sst, because snapshot 8894878975848 already exists.
09:59:59.758 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000199.sst, because snapshot 8894878975848 already exists.
09:59:59.760 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000200.sst, because snapshot 8894878975848 already exists.
09:59:59.761 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000201.sst, because snapshot 8894878975848 already exists.
09:59:59.763 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000202.sst, because snapshot 8894878975848 already exists.
09:59:59.767 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000203.sst, because snapshot 8894878975848 already exists.
09:59:59.769 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000204.sst, because snapshot 8894878975848 already exists.
09:59:59.770 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk 000205.sst, because snapshot 8894878975848 already exists.
09:59:59.771 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk CURRENT, because snapshot 8894878975848 already exists.
09:59:59.772 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk MANIFEST-000004, because snapshot 8894878975848 already exists.
09:59:59.773 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Ignore snapshot chunk OPTIONS-000084, because snapshot 8894878975848 already exists.
09:59:59.773 [] [log-restore-7] ERROR io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService - Unexpected error occurred when moving snapshot 8894878975848 to /usr/local/zeebe/data/partition-7/state/snapshots/8894878975848
java.nio.file.NoSuchFileException: /usr/local/zeebe/data/partition-7/state/snapshots/8894878975848-restore-log
        at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86) ~[?:1.8.0_212]
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102) ~[?:1.8.0_212]
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107) ~[?:1.8.0_212]
        at sun.nio.fs.UnixCopyFile.move(UnixCopyFile.java:409) ~[?:1.8.0_212]
        at sun.nio.fs.UnixFileSystemProvider.move(UnixFileSystemProvider.java:262) ~[?:1.8.0_212]
        at java.nio.file.Files.move(Files.java:1395) ~[?:1.8.0_212]
        at io.zeebe.logstreams.state.FileSnapshotConsumer.moveValidSnapshot(FileSnapshotConsumer.java:119) ~[zeebe-logstreams-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.logstreams.state.FileSnapshotConsumer.completeSnapshot(FileSnapshotConsumer.java:45) ~[zeebe-logstreams-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.distributedlog.restore.snapshot.RestoreSnapshotReplicator.lambda$restoreInternal$0(RestoreSnapshotReplicator.java:107) ~[zeebe-logstreams-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760) ~[?:1.8.0_212]
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736) ~[?:1.8.0_212]
        at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:442) ~[?:1.8.0_212]
        at io.atomix.utils.concurrent.SingleThreadContext$1.lambda$execute$0(SingleThreadContext.java:53) ~[atomix-utils-3.2.0-alpha2.jar:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_212]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_212]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:1.8.0_212]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:1.8.0_212]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_212]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_212]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_212]
09:59:59.774 [] [log-restore-7] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService-7 - Snapshot restore failed 8894878975848
java.lang.RuntimeException: Could not restore snapshot 8894878975848. Failed to move valid snapshot.
        at io.zeebe.distributedlog.restore.snapshot.RestoreSnapshotReplicator.lambda$restoreInternal$0(RestoreSnapshotReplicator.java:115) ~[zeebe-logstreams-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760) ~[?:1.8.0_212]
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736) ~[?:1.8.0_212]
        at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:442) ~[?:1.8.0_212]
        at io.atomix.utils.concurrent.SingleThreadContext$1.lambda$execute$0(SingleThreadContext.java:53) ~[atomix-utils-3.2.0-alpha2.jar:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_212]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_212]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:1.8.0_212]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:1.8.0_212]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_212]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_212]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_212]
09:59:59.774 [] [raft-server-raft-atomix-partition-7-state] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService-7 - Restoring local log failed at position 7984346530400, retrying.
java.util.concurrent.CompletionException: java.lang.RuntimeException: Could not restore snapshot 8894878975848. Failed to move valid snapshot.
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292) ~[?:1.8.0_212]
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308) ~[?:1.8.0_212]
        at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:943) ~[?:1.8.0_212]
        at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:926) ~[?:1.8.0_212]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474) ~[?:1.8.0_212]
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977) ~[?:1.8.0_212]
        at io.zeebe.distributedlog.restore.snapshot.RestoreSnapshotReplicator.failReplication(RestoreSnapshotReplicator.java:124) ~[zeebe-logstreams-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.distributedlog.restore.snapshot.RestoreSnapshotReplicator.lambda$restoreInternal$0(RestoreSnapshotReplicator.java:111) ~[zeebe-logstreams-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760) ~[?:1.8.0_212]
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736) ~[?:1.8.0_212]
        at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:442) ~[?:1.8.0_212]
        at io.atomix.utils.concurrent.SingleThreadContext$1.lambda$execute$0(SingleThreadContext.java:53) ~[atomix-utils-3.2.0-alpha2.jar:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_212]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_212]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:1.8.0_212]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:1.8.0_212]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_212]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_212]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_212]
Caused by: java.lang.RuntimeException: Could not restore snapshot 8894878975848. Failed to move valid snapshot.
        at io.zeebe.distributedlog.restore.snapshot.RestoreSnapshotReplicator.lambda$restoreInternal$0(RestoreSnapshotReplicator.java:115) ~[zeebe-logstreams-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        ... 11 more
09:59:59.774 [] [raft-server-raft-atomix-partition-7-state] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService-7 - Restoring local log from position 7984346530400 to 8035885574312
09:59:59.775 [] [netty-messaging-event-epoll-client-5] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService-7 - Restoring snapshot 8894878975848 from server 2 (expecting 45 chunks)
```

--- a/logstreams/src/main/java/io/zeebe/distributedlog/restore/snapshot/RestoreSnapshotReplicator.java
+++ b/logstreams/src/main/java/io/zeebe/distributedlog/restore/snapshot/RestoreSnapshotReplicator.java
@@ -64,7 +64,13 @@ public class RestoreSnapshotReplicator {
       MemberId server, long snapshotId, int numChunks) {
     this.numChunks = numChunks;
     final CompletableFuture<Tuple<Long, Long>> result = new CompletableFuture<>();
-    restoreInternal(server, snapshotId, 0, result);
+
+    if (restoreContext.getStateStorage().existSnapshot(snapshotId)) {
+      result.complete(restoreContext.getSnapshotPositionSupplier().get());
+    } else {
+      restoreInternal(server, snapshotId, 0, result);
+    }
+
     return result;
   }
 
@@ -105,7 +111,8 @@ public class RestoreSnapshotReplicator {
               }
 
               if (snapshotConsumer.completeSnapshot(snapshotId)) {
-                final Tuple positions = restoreContext.getSnapshotPositionSupplier().get();
+                final Tuple<Long, Long> positions =
+                    restoreContext.getSnapshotPositionSupplier().get();
                 future.complete(positions);
               } else {
                 failReplication(

--- a/logstreams/src/test/java/io/zeebe/distributedlog/restore/snapshot/RestoreSnapshotReplicatorTest.java
+++ b/logstreams/src/test/java/io/zeebe/distributedlog/restore/snapshot/RestoreSnapshotReplicatorTest.java
@@ -17,6 +17,8 @@ package io.zeebe.distributedlog.restore.snapshot;
 
 import static io.zeebe.test.util.TestUtil.waitUntil;
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
 
 import io.atomix.cluster.MemberId;
 import io.zeebe.distributedlog.restore.impl.ControllableRestoreClient;
@@ -24,26 +26,38 @@ import io.zeebe.distributedlog.restore.impl.ControllableSnapshotRestoreContext;
 import io.zeebe.distributedlog.restore.snapshot.impl.InvalidSnapshotRestoreResponse;
 import io.zeebe.distributedlog.restore.snapshot.impl.SuccessSnapshotRestoreResponse;
 import io.zeebe.logstreams.state.SnapshotChunk;
+import io.zeebe.logstreams.state.StateStorage;
 import io.zeebe.util.collection.Tuple;
 import java.util.concurrent.CompletableFuture;
 import org.junit.Before;
+import org.junit.Rule;
 import org.junit.Test;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+import org.mockito.quality.Strictness;
 
 public class RestoreSnapshotReplicatorTest {
-  private ControllableRestoreClient client = new ControllableRestoreClient();
-  private ControllableSnapshotRestoreContext restoreContext =
+  private final ControllableRestoreClient client = new ControllableRestoreClient();
+  private final ControllableSnapshotRestoreContext restoreContext =
       new ControllableSnapshotRestoreContext();
-  private RecordingSnapshotConsumer snapshotConsumer = new RecordingSnapshotConsumer();
-  private RestoreSnapshotReplicator snapshotReplicator =
+  private final RecordingSnapshotConsumer snapshotConsumer = new RecordingSnapshotConsumer();
+  private final RestoreSnapshotReplicator snapshotReplicator =
       new RestoreSnapshotReplicator(client, restoreContext, snapshotConsumer, Runnable::run);
-  private MemberId server = MemberId.anonymous();
+  private final MemberId server = MemberId.anonymous();
   private final ControllableSnapshotChunk responseChunk = new ControllableSnapshotChunk();
+
+  @Rule
+  public final MockitoRule mockitoRule = MockitoJUnit.rule().strictness(Strictness.STRICT_STUBS);
+
+  @Mock private StateStorage stateStorage = mock(StateStorage.class);
 
   @Before
   public void setup() {
     client.reset();
     restoreContext.reset();
     snapshotConsumer.reset();
+    restoreContext.setProcessorStateStorage(stateStorage);
   }
 
   @Test
@@ -53,7 +67,7 @@ public class RestoreSnapshotReplicatorTest {
     final long snapshotId = 10;
     final int numChunks = 5;
 
-    restoreContext.setPositionSupplier(() -> new Tuple(5L, 10L));
+    restoreContext.setPositionSupplier(() -> new Tuple<>(5L, 10L));
 
     for (int i = 0; i < numChunks; i++) {
       client.completeRequestSnapshotChunk(
@@ -79,7 +93,7 @@ public class RestoreSnapshotReplicatorTest {
     final long snapshotId = 10;
     final int numChunks = 5;
 
-    restoreContext.setPositionSupplier(() -> new Tuple(5L, 10L));
+    restoreContext.setPositionSupplier(() -> new Tuple<>(5L, 10L));
 
     client.completeRequestSnapshotChunk(0, new InvalidSnapshotRestoreResponse());
 
@@ -97,7 +111,7 @@ public class RestoreSnapshotReplicatorTest {
     final long snapshotId = 10;
     final int numChunks = 5;
 
-    restoreContext.setPositionSupplier(() -> new Tuple(5L, 10L));
+    restoreContext.setPositionSupplier(() -> new Tuple<>(5L, 10L));
 
     final CompletableFuture<Tuple<Long, Long>> replicate =
         snapshotReplicator.restore(server, snapshotId, numChunks);
@@ -113,9 +127,25 @@ public class RestoreSnapshotReplicatorTest {
     client.completeRequestSnapshotChunk(2, new RuntimeException());
 
     // then
-    waitUntil(() -> replicate.isDone());
+    waitUntil(replicate::isDone);
     assertThat(replicate).isCompletedExceptionally();
     assertThat(snapshotConsumer.getConsumedChunks().size()).isEqualTo(0);
+  }
+
+  @Test
+  public void shouldCompleteImmediatelyIfSnapshotAlreadyExists() {
+    // given
+    final long snapshotId = 10;
+    final Tuple<Long, Long> positions = new Tuple<>(5L, 10L);
+    restoreContext.setPositionSupplier(() -> positions);
+    when(stateStorage.existSnapshot(snapshotId)).thenReturn(true);
+
+    // when
+    final CompletableFuture<Tuple<Long, Long>> result =
+        snapshotReplicator.restore(server, snapshotId, 5);
+
+    // then
+    assertThat(result).isCompletedWithValue(positions);
   }
 
   private static final class ControllableSnapshotChunk implements SnapshotChunk {

